language/vm/src/deserializer.rs

```diff
@@ -661,44 +661,117 @@ fn load_signature_tokens(cursor: &mut Cursor<&[u8]>) -> BinaryLoaderResult<Vec<S
 
 /// Deserializes a `SignatureToken`.
 fn load_signature_token(cursor: &mut Cursor<&[u8]>) -> BinaryLoaderResult<SignatureToken> {
-    if let Ok(byte) = cursor.read_u8() {
-        match SerializedType::from_u8(byte)? {
-            SerializedType::BOOL => Ok(SignatureToken::Bool),
-            SerializedType::U8 => Ok(SignatureToken::U8),
-            SerializedType::U64 => Ok(SignatureToken::U64),
-            SerializedType::U128 => Ok(SignatureToken::U128),
-            SerializedType::ADDRESS => Ok(SignatureToken::Address),
-            SerializedType::VECTOR => {
-                let ty = load_signature_token(cursor)?;
-                Ok(SignatureToken::Vector(Box::new(ty)))
-            }
-            SerializedType::REFERENCE => {
-                let ref_token = load_signature_token(cursor)?;
-                Ok(SignatureToken::Reference(Box::new(ref_token)))
-            }
-            SerializedType::MUTABLE_REFERENCE => {
-                let ref_token = load_signature_token(cursor)?;
-                Ok(SignatureToken::MutableReference(Box::new(ref_token)))
+    use SerializedType as S;
+
+    enum TypeConstructor {
```

Comment:
Maybe `TypeBuilder` or `SignatureTokenBuilder`? 

I find the term `TypeConstructor` a bit confusing for the way that it is working in the `StructInst` case, but I could be convinced otherwise :) Just a thought

Reply:
to me it feels the `TypeConstructor` is something for the struct instantiation and maybe better used in that context only. But I have to admit that it may very well lead to something exactly like this and I am fine with this going in!

Reply:
Well that's exactly what the type constructor does if you allow partial application and currying :) But yeah I agree we should try not to introduce such concepts as Move doesn't have them.